cmd/node_agent/na/nodeagent.go

```diff
@@ -148,57 +206,47 @@ func (na *nodeAgentInternal) Start() {
 	}
 }
 
-func (na *nodeAgentInternal) createRequest() ([]byte, *pb.Request) {
+func (na *nodeAgentInternal) createRequest() ([]byte, *pb.Request, error) {
 	csr, privKey, err := ca.GenCSR(ca.CertOptions{
 		Host:       na.config.ServiceIdentity,
 		Org:        na.config.ServiceIdentityOrg,
 		RSAKeySize: na.config.RSAKeySize,
 	})
 
 	if err != nil {
-		glog.Fatalf("Failed to generate CSR: %v", err)
+		return nil, nil, fmt.Errorf("failed to generate CSR: %v", err)
 	}
 
-	return privKey, &pb.Request{
-		CsrPem: csr,
-	}
+	return privKey, &pb.Request{CsrPem: csr}, nil
 }
 
-func (na *nodeAgentInternal) sendCSR() ([]byte, *pb.Response, error) {
-	glog.Infof("Sending out CSR to CA...")
-	dialOptions, err := na.pr.GetDialOptions(na.config)
-	if err != nil {
-		glog.Errorf("Cannot construct the dial options with error %v", err)
-		return nil, nil, err
+func (na *nodeAgentInternal) getWaitTimeFromCert(
+	certBytes []byte, now time.Time, gracePeriodPercentage int) (time.Duration, error) {
+	cert, certErr := pki.ParsePemEncodedCertificate(certBytes)
+	if certErr != nil {
+		return time.Duration(0), certErr
 	}
-	conn, err := grpc.Dial(na.config.IstioCAAddress, dialOptions...)
-	if err != nil {
-		glog.Fatalf("Failed to dial %s: %v", na.config.IstioCAAddress, err)
+	timeToExpire := cert.NotAfter.Sub(now)
+	if timeToExpire < 0 {
+		return time.Duration(0), fmt.Errorf("certificate already expired at %s, but now is %s",
+			cert.NotAfter, now)
 	}
-
-	defer func() {
-		if closeErr := conn.Close(); closeErr != nil {
-			glog.Fatalf("Failed to close connection")
-		}
-	}()
-
-	client := pb.NewIstioCAServiceClient(conn)
-	privKey, req := na.createRequest()
-	resp, err := client.HandleCSR(context.Background(), req)
-	if err != nil {
-		glog.Errorf("CSR request failed %v", err)
-		return nil, nil, err
+	gracePeriod := cert.NotAfter.Sub(cert.NotBefore) * time.Duration(gracePeriodPercentage) / time.Duration(100)
+	// Wait until the grace period starts.
+	waitTime := timeToExpire - gracePeriod
+	if waitTime < 0 {
+		waitTime = 0
 	}
-
-	return privKey, resp, nil
+	return waitTime, nil
 }
 
-func (na *nodeAgentInternal) writeToFile(privKey []byte, cert []byte) {
+func (na *nodeAgentInternal) writeToFile(privKey []byte, cert []byte) error {
```

Comment:
I personally think we should have all the file I/O functions in a util/ folder. This reduces duplication of file I/O functions and facilitates our tests. I can refactor it if you think necessary. I think for now, we can temporarily keep it here.